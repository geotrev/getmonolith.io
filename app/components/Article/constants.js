import { Tooltips, Accordions, Collapsibles, Modals, Dropdowns } from "undernet"

export const COMPONENTS = [Tooltips, Accordions, Collapsibles, Modals, Dropdowns]
