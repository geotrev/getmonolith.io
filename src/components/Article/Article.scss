@import "../../global";

.article-wrapper {
  pre {
    margin-bottom: 32px !important;
  }

  .filler {
    background: $gray900;
    border-radius: $global-border-radius;
    margin-bottom: 1rem;

    &.has-padding {
      padding: 1rem;
    }
  }

  h2,
  h3 {
    margin-top: 54px;
  }

  .filler-bg {
    background: $gray900;
    padding: 1rem;
    margin-bottom: 1rem;
    border-radius: $global-border-radius;

    .filler {
      background: $gray800;
      color: white;
      margin-bottom: 0;
    }
  }
}
