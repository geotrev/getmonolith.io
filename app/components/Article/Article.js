import React, { useState, useEffect } from "react"
import Markdown from "react-markdown"
import Prism from "prismjs"
import classNames from "classnames"
import PropTypes from "prop-types"

import Undernet from "undernet"
import { COMPONENTS } from "./constants"
import ScrollUpOnMount from "app/components/ScrollUpOnMount"

export default function Article(props) {
  const [domIsLoaded, setDomIsLoaded] = useState(false)

  const componentUnmountFunction = () => {
    COMPONENTS.forEach(component => Undernet[component].stop())
  }

  useEffect(() => {
    Prism.highlightAll()
    COMPONENTS.forEach(component => Undernet[component].start())
    setDomIsLoaded(true)

    return componentUnmountFunction
  }, [])

  return (
    <article
      className={classNames("article-wrapper has-no-padding column", {
        fadeIn: domIsLoaded,
      })}
    >
      <ScrollUpOnMount />
      <Markdown source={props.children} escapeHtml={false} />
    </article>
  )
}

Article.propTypes = {
  children: PropTypes.any,
}
