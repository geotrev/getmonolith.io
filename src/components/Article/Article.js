import React, { Component } from "react"
import "./Article.scss"

import Markdown from "react-markdown"
import Prism from "prismjs"
import Undernet from "undernet"

export default class Article extends Component {
  constructor() {
    super()
  }

  componentPage() {
    const components = ["/modals", "/accordions"]

    components.map(urlStr => {
      if (window.location.href.includes(urlStr)) {
        return true
      }
    })

    return false
  }

  componentDidMount() {
    if (this.componentPage()) {
      Undernet.start()
    }

    Prism.highlightAll()
  }

  componentWillUnmount() {
    if (this.componentPage()) {
      Undernet.stop()
    }
  }

  render() {
    return (
      <article className="article-wrapper has-no-padding column">
        <Markdown source={this.props.children} escapeHtml={false} />
      </article>
    )
  }
}
