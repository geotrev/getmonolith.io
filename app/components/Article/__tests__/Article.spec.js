import React from "react"
import Article from "../Article"
import Undernet from "undernet"
import Prism from "prismjs"

global.scrollTo = jest.fn()

jest.mock("undernet", () => ({
  start: jest.fn(),
  stop: jest.fn(),
}))

jest.mock("prismjs", () => ({
  highlightAll: jest.fn(),
}))

describe("<Article />", () => {
  let wrapper
  beforeEach(() => {
    const md = "# Test header \n So neat"
    wrapper = mount(<Article>{md}</Article>)
  })

  it("renders", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("calls window.scrollTo on mount", () => {
    expect(window.scrollTo).toHaveBeenCalledWith(0, 0)
  })

  it("calls Prism.highlightAll on mount", () => {
    expect(Prism.highlightAll).toHaveBeenCalled()
  })

  it("starts Undernet on mount", () => {
    expect(Undernet.start).toHaveBeenCalled()
  })

  it("stops Undernet on unmount", () => {
    wrapper.unmount()
    expect(Undernet.stop).toHaveBeenCalled()
  })

  it("sets fadeIn class on wrapper on mount", () => {
    expect(
      wrapper
        .find("article")
        .prop("className")
        .includes("fadeIn")
    ).toBe(true)
  })
})
