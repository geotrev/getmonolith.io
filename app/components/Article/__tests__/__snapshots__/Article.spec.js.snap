// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> #render renders 1`] = `
<Article>
  <article
    className="article-wrapper has-no-p column fadeIn"
  >
    <ScrollUpOnMount />
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={false}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="# Test header 
 So neat"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1"
      >
        <Heading
          key="heading-1-1"
          level={1}
        >
          <h1>
            <TextRenderer
              key="text-1-3"
              nodeKey="text-1-3"
              value="Test header"
            >
              Test header
            </TextRenderer>
          </h1>
        </Heading>
        <p
          key="paragraph-2-1"
        >
          <TextRenderer
            key="text-2-1"
            nodeKey="text-2-1"
            value=" So neat"
          >
             So neat
          </TextRenderer>
        </p>
      </Root>
    </ReactMarkdown>
  </article>
</Article>
`;
